### PR TITLE
community[patch]: Improve import callbacks to make it IDE friendly

### PR DESCRIPTION
**Pull request by eyurtsev**
* declares __all__ as a list of strings (instead of dynamically computing it)
* import type definitions when TYPE_CHECKING is true

### CODE REVIEW

**Comment by pierreveron** [comment]
@eyurtsev This solves my [issue](https://github.com/langchain-ai/langchain/pull/18928#issuecomment-2037464828) too in VSCode but a new one appeared. Pylint complains about the imports not being exported despite the `__all__ `variable.

image

My guess is that the linter doesn't compute the list from the dict. My fix to that is to put back the original list.
```python
__all__ = [
    "AimCallbackHandler",
    "ArgillaCallbackHandler",
    "ArizeCallbackHandler",
    "PromptLayerCallbackHandler",
    "ArthurCallbackHandler",
    "ClearMLCallbackHandler",
    "CometCallbackHandler",
    "ContextCallbackHandler",
    "HumanApprovalCallbackHandler",
    "InfinoCallbackHandler",
    "MlflowCallbackHandler",
    "LLMonitorCallbackHandler",
    "OpenAICallbackHandler",
    "LLMThoughtLabeler",
    "StreamlitCallbackHandler",
    "WandbCallbackHandler",
    "WhyLabsCallbackHandler",
    "get_openai_callback",
    "wandb_tracing_enabled",
    "FlyteCallbackHandler",
    "SageMakerCallbackHandler",
    "LabelStudioCallbackHandler",
    "TrubricsCallbackHandler",
    "FiddlerCallbackHandler",
]
```

**Comment by eyurtsev** [comment]
@pierreveron got it -- it sounds like we should not compute the __all__ list dynamically and instead rely on unit tests to catch any drift

**Comment by eyurtsev** [comment]
@pierreveron made required changes, we'll merge and please confirm that it all works as expected

**Comment by pierreveron** [comment]
> @pierreveron made required changes, we'll merge and please confirm that it all works as expected

Super cool thanks @eyurtsev. I think you should do the same for all modules in `langchain_community` that were changed by the [PR](https://github.com/langchain-ai/langchain/pull/18928). There can be cascading errors, for example when importing `from langchain.memory import ChatMessageHistory` as `ChatMessageHistory` export is not caught by the linter in `libs/community/langchain_community/chat_message_histories/__init__.py`

File: libs/community/tests/unit_tests/callbacks/test_imports.py [hunk 1]
```diff
@@ -29,4 +29,6 @@
 
 
 def test_all_imports() -> None:
+    """Test that __all__ is correctly set."""
     assert set(__all__) == set(EXPECTED_ALL)
```
Comment (leo-gan) [inline libs/community/tests/unit_tests/callbacks/test_imports.py:33]: I guess, we don't need the EXPECTED_ALL anymore, no?

**Review by baskaryan — APPROVED** [review]

File: libs/community/tests/unit_tests/callbacks/test_imports.py (hunk 1 above)
Comment (eyurtsev) [reply libs/community/tests/unit_tests/callbacks/test_imports.py:33]: We usually maintain a second list of hard-coded imports in unit tests to catch accidental breaking changes (e.g., removal of an import)

Comment (eyurtsev) [reply libs/community/tests/unit_tests/callbacks/test_imports.py:33]: We could remove in this case and just make sure that set(__all__) == the keys from the module. I don't have a strong preference

**Comment by leo-gan** [comment]
@eyurtsev I'll refactor other `langchain_community` modules whenever applicable.

**Pull request merged**